Xarray refactor: intro raster and intro xarray (#3)
jorisvandenbossche/course-python-geospatial: notebooks/11-xarray-intro.md

```{code-cell} ipython3
import numpy as np
```

We could also read the GeoTiff data directly into a numpy array using `rasterio`:

```{code-cell} ipython3
import rasterio

with rasterio.open(data_file) as src:
    herstappe_array = src.read()
    herstappe_meta = src.meta
```

```{code-cell} ipython3
herstappe_array
```

As we learnt in the previous lesson, Rasterio returns a Numpy `ndarray`:

```{code-cell} ipython3
type(herstappe_array)
```

Numpy supports different `dtype`s (`float`, `int`,...), but all elements of an array do have the same dtype. Note that NumPy auto-detects the data-type from the input.

```{code-cell} ipython3
herstappe_array.dtype
```

The data type of this specific array `herstappe_array` is float32. More information on the data types Numpy supports is available in the [documentation](https://numpy.org/devdocs/user/basics.types.html#array-types-and-conversions-between-types).

+++

> Converting to another data type is supported by `astype` method. When floats are preferred during calculation:

Reviewer: We showed astype previously in the numpy section, but we still use it in the exercises. Maybe keep a mention of it?

Author: ok, I added a small section on it.

> Or just rename the coordinate names itself:

Reviewer: ```suggestion
Or just rename the dimension names itself:
```

Or not .. In the overview list we in the previous notebook (it might be worth repeating that in this notebook) we call it dimensions. 

I think typically xarray "dimensions" for the names of the dimensions, and "coordinates" for the actual coordinate values for a specific dimension (which is of course not a clear cut line).

Author: I would keep coordinate here, the docstring of `rename` says also coordinate

> Returns a new DataArray with renamed coordinates or a new name.

I see these are 'dimension coordinates' (whereas you can also have non-dimension coordinates), so these are the names of the coordinates linked to the dimension. A dimension does not have a name on itself... _I'm realizing this while typing, so not sure if I did this consistently in the course notes_.

Reviewer: > A dimension does not have a name on itself

I think a dimension actually always has a name, but doesn't necessarily have coordinates (and indeed coordinates are not always linked to a dimension). If you create a DataArray from a numpy array without specifying any dims/coords, you get "dim_0" etc for the dimension names (without coordinates).

> herstappe.rename({"x": "longitude", "y": "latitude"})

Reviewer: It's a nice example to rename this, but the values are not longitude/latitude (since that would mean geographical coordinates, and it are projected coordinates), so the new name is actually incorrect ..

Author: Good point! I'll make it easting/northing.

> These __name dimensions__ can be used to extract (slice) data using these names making data selections very declarative:

Reviewer: ```suggestion
These __named dimensions__ can be used to extract (slice) data making data selections very declarative:
```

Or just rename the coordinate names itself:

```{code-cell} ipython3
herstappe.rename({"x": "easting", "y": "northing"})
```

These __dimension coordinate names__ can be used to extract (slice) data making data selections very declarative:

```{code-cell} ipython3
herstappe.sel(band='red')
```

The data type of this `xarray.DataArray` `herstappe` is 'float32'. Xarray uses the data types provided by Numpy. More information on the data types Numpy supports is available in the [documentation](https://numpy.org/devdocs/user/basics.types.html#array-types-and-conversions-between-types).

Converting to another data type is supported by `astype()` method:

```{code-cell} ipython3
herstappe.astype('float64')   # .nbytes
```

Using xarray:

- Data stored as a Numpy arrays.

> - Dimensions do have a coordinate with a name.

Reviewer: ```suggestion
- Dimensions do have a name.
```

(restoring the original text here, which I think is strictly speaking more correct, and coordinates are explained in the next bullet point)

> herstappe[0]

Reviewer: Just above, it still uses "data_file", but you renamed it to "herstappe_file"

Author: fixed

> fig.colorbar(ll);

Reviewer: TODO for if there is time: it might be good to include some more instructions + hints to be able to create those figures

Author: I adjusted the exercise with some more background and detailed steps on how to apply the normalization.
